numerics/quadrature_body.hpp

```diff
@@ -24,35 +34,133 @@ Primitive<std::invoke_result_t<Function, Argument>, Argument> Gauss(
+  // We look for an estimated relative error smaller than
+  // |relative_tolerance * points|: since the integral is computed from |points|
+  // evaluations of |f|, it will necessarily carry a relative error proportional
+  // to |points|, so it makes no sense to look for convergence beyond that.
+  if (relative_error_estimate > relative_tolerance * points) {
+    if constexpr (points > 1 << 24) {
+      LOG(FATAL) << "Too many refinements while integrating from "
+                 << lower_bound << " to " << upper_bound;
+    } else {
+      return AutomaticClenshawCurtis<points + points - 1>(
+          f, lower_bound, upper_bound, relative_tolerance, estimate);
+    }
+  }
+  return estimate;
+}
+
+template<int initial_points, typename Argument, typename Function>
+Primitive<std::invoke_result_t<Function, Argument>, Argument>
+AutomaticClenshawCurtis(
+    Function const& f,
+    Argument const& lower_bound,
+    Argument const& upper_bound,
+    double const relative_tolerance) {
+  using Result = Primitive<std::invoke_result_t<Function, Argument>, Argument>;
+  // TODO(egg): factor the evaluations of f.
+  Result const estimate =
+      ClenshawCurtis<initial_points>(f, lower_bound, upper_bound);
+  return AutomaticClenshawCurtis<initial_points + initial_points - 1>(
+      f,
+      lower_bound,
+      upper_bound,
+      relative_tolerance,
+      estimate);
+}
+
+template<int points, typename Argument, typename Function>
+Primitive<std::invoke_result_t<Function, Argument>, Argument> ClenshawCurtis(
+    Function const& f,
+    Argument const& lower_bound,
+    Argument const& upper_bound) {
+  // We follow the notation from [Gen72b] and [Gen72c].
+  using Value = std::invoke_result_t<Function, Argument>;
+
+  constexpr int N = points - 1;
+  constexpr int log2_N = FloorLog2(N);
+  static_assert(N == 1 << log2_N);
+
+  Difference<Argument> const half_width = (upper_bound - lower_bound) / 2;
+
+  constexpr Angle N⁻¹π = π * Radian / N;
+
+  // We use a discrete Fourier transform rather than a cosine transform, see
+  // [Gen72c], (3).
+  // TODO(egg): Consider a discrete cosine transform, ideally incrementally
+  // computed to improve the performance of the automatic quadrature.
+
+  // If we identify [lower_bound, upper_bound] with [-1, 1],
+  // f_cos_N⁻¹π[s] is f(cos πs/N).
+  std::vector<Value> f_cos_N⁻¹π;
+  f_cos_N⁻¹π.resize(2 * N);
+  for (int s = 0; s <= N; ++s) {
+    // The N + 1 evaluations.
+    f_cos_N⁻¹π[s] = f(lower_bound + half_width * (1 + Cos(N⁻¹π * s)));
+  }
+  for (int s = N + 1; s <= 2 * N - 1; ++s) {
+    f_cos_N⁻¹π[s] = f_cos_N⁻¹π[2 * N - s];  // [Gen72c] (5).
+  }
+
+  auto const fft = std::make_unique<FastFourierTransform<Value, Angle, 2 * N>>(
+      f_cos_N⁻¹π, N⁻¹π);
+  auto const& a = *fft;
+
+  // [Gen72b] (7), factoring out the division by N.
+  Value Σʺ{};
+  for (std::int64_t n = 0; n <= N; n += 2) {
+    // The notation g is from [OLBC10], 3.5.17.
+    int gₙ = n == 0 || n == N ? 1 : 2;
```

Comment:
`const`

Note that the `?:` is likely to involve a jump.  A slightly faster way to write it would be to extract `n = 0` and `n = N` from the loop, and loop from `n = 1` to `n < N`.

Reply:
Shrug.